pytests/test_debugging.py

```diff
@@ -0,0 +1,504 @@
+
+    """
+    Debugger: step out of the current stack frame.
+    """
+    def step_out(self) -> [dict]:
+        return self.send_message(DebuggerClientMessage.DEBUGGER_STEP_OUT)
+
+    """
+    Debugger: step over the next instruction.
+    """
+    def step_over(self) -> [dict]:
+        return self.send_message(DebuggerClientMessage.DEBUGGER_STEP_OVER)
+
+
+def deploy_script(env: Env):
+    env.expect('TFUNCTION', 'DELETE', 'foo')
+    env.expect('TFUNCTION', 'LOAD', 'debug', SCRIPT).contains('The V8 remote debugging server is waiting for a connection')
+
+@gearsTest()
+def testDeployConnectStartDisconnect(env: Env):
+    deploy_script(env)
+    client = DebuggerClient(env)
+    client.disconnect()
+
+@gearsTest()
+def testSetBreakpointAndHitIt(env: Env):
+    deploy_script(env)
+    client = DebuggerClient(env)
+    breakpoint = client.set_breakpoint(2, 1)
+    env.assertEqual(len(client.breakpoints), 1)
+    # Now proceed from the very first instruction to the very first
+    # breakpoint we've just set.
+    client.resume_and_wait_until_pauses()
+    env.assertIsNotNone(client.pause)
+    env.assertContains("callFrames", client.pause["params"])
+    # Check that we hit the breakpoint we set above.
+    env.assertEqual(breakpoint.get_id(), client.pause["params"]["hitBreakpoints"][0])
+
+    client.disconnect()
+
+@gearsTest()
+def testSetBreakpointAndRemoveIt(env: Env):
+    deploy_script(env)
+    client = DebuggerClient(env)
+    breakpoint = client.set_breakpoint(2, 1)
+    env.assertEqual(len(client.breakpoints), 1)
+    breakpoint.remove()
+    # Now proceed from the very first instruction till the end.
+    env.assertEqual({}, client.resume()[0]["result"])
+    # The next message should be a "resumed" event from the server.
+    resumed_event = client.receive_event()
+    env.assertEqual("Debugger.resumed", resumed_event["method"])
+    # Now wait for the possible pause event to happen. It should not
+    # happen now, for we have no breakpoints set, and the script
+    # execution should just continue with no problem. We wait for
+    # five seconds to let *any* sort of execution to continue on any VM,
+    # so to avoid the flakiness of the test result.
+    try:
+        env.assertIsNone(client.receive_event(5.0))
+    except TimeoutError:
+        # Receiving a timeout error means there was no debugger pause,
+        # exactly what we expect in this case.
+        pass
+    except BaseException as e:
+        # We should not receive, however, any other sorts of error.
+        raise(e)
+
+    client.disconnect()
+
+"""
+The debugging is considered disabled, when the debug server
+address argument is not correctly specified.
+"""
+@gearsTest(debugServerAddress="")
+def testNotPossibleToDebugWhenDebuggingIsDisabled(env: Env):
+    env.expect('TFUNCTION', 'DELETE', 'foo')
+    env.expect('TFUNCTION', 'LOAD', 'debug', SCRIPT).contains('The V8 remote debugging server is waiting for a connection')
```

Comment:
Shouldn't we expect here a message that indicating that debugger is disabled?

Reply:
Also the default should be disable, so how come the other tests does not need to enable it?

Reply:
Guess, my eyes were dry already, indeed a huge overlook. Let me see.